Sources/Penny/MainService/PennyService.swift

```diff
@@ -147,30 +164,59 @@ struct PennyService: MainService {
             swiftReleasesChecker: swiftReleasesChecker,
             reactionCache: reactionCache
         )
-        let context = HandlerContext(
-            services: services,
-            botStateManager: BotStateManager(services: services)
-        )
+        context.botStateManager = BotStateManager(context: context)
+        context.discordEventListener = DiscordEventListener(bot: bot, context: context)
 
         await CommandsManager(context: context).registerCommands()
 
         return context
     }
 
-    func afterConnectCall(context: HandlerContext) async throws {
-        /// Wait 5 seconds to make sure the bot is completely connected to Discord through websocket,
-        /// and so it can receive events already.
-        /// This is here until when/if DiscordBM gains better support for notifying you
-        /// of the first connection.
-        /// We could manually handle that here too, but i'd like it to be available in DiscordBM.
-        try await Task.sleep(for: .seconds(5))
-        /// Initialize `BotStateManager` after `bot.connect()` and `bot.makeEventsStream()`.
-        /// since it communicates through Discord and will need the Gateway connection.
-        await context.botStateManager.start {
-            /// These contain cached stuff and need to wait for `BotStateManager`.
-            context.services.evolutionChecker.run()
-            context.services.soChecker.run()
-            context.services.swiftReleasesChecker.run()
-        }
+    func runServices(context: HandlerContext) async throws {
+        /// Services that need to wait for bot connection
+        let botStateManagerWrappedService = WaiterService(
+            underlyingService: context.botStateManager,
+            processingOn: context.backgroundProcessor,
+            passingContinuationWith: {
+                await context.discordEventListener.addConnectionWaiterContinuation($0)
+            }
+        )
+
+        /// Services that need to wait for caches population
+        let evolutionCheckerWrappedService = WaiterService(
+            underlyingService: context.evolutionChecker,
+            processingOn: context.backgroundProcessor,
+            passingContinuationWith: {
+                await context.botStateManager.addCachesPopulationContinuation($0)
+            }
+        )
+        let soCheckerWrappedService = WaiterService(
+            underlyingService: context.soChecker,
+            processingOn: context.backgroundProcessor,
+            passingContinuationWith: {
+                await context.botStateManager.addCachesPopulationContinuation($0)
+            }
+        )
+        let swiftReleasesCheckerWrappedService = WaiterService(
+            underlyingService: context.swiftReleasesChecker,
+            processingOn: context.backgroundProcessor,
+            passingContinuationWith: {
+                await context.botStateManager.addCachesPopulationContinuation($0)
+            }
+        )
```

Comment:
See comments as to why we're using "wrapped" services.
The `WaiterService` type also has some doc comments.